### PR TITLE
[RISC-V] LoongArch code sync

### PR DESCRIPTION
**Pull request by tomeksowi**
Changes ported from LoongArch useful to RISC-V.

* Sign-extend index in genRangeCheck
Analogous fix to #86435 on Loongarch, fixes System.Linq.Expressions.Tests.RuntimeVariablesTests.MixedScope failing with IndexOutOfRangeException.
* Encode stack pointer as 1 in gcinfo
Change similar to #85092. Sp(x2) was encoded as 2^8=10, which is 4 bits long and the encoding won't fit in a single chunk of STACK_BASE_REGISTER_ENCBASE=2 bits.
* #ifdef DEBUG entire if (needCheckOv) section since assert are only active in debug mode
* Use REG_DEFAULT_HELPER_CALL_TARGET instead of REG_T2 to improve readability
* Use largeframe condition like on LoongArch

Part of #84834
cc @wscho77 @HJLeee @clamp03 @JongHeonChoi @t-mustafin @gbalykov @viewizard @ashaurtaev @brucehoult @sirntar @yurai007 @bajtazar

### CODE REVIEW

**Comment by DragosDanielBoia** [comment]
@dotnet-policy-service rerun

**Review by clamp03 — APPROVED** [review]

**Comment by clamp03** [comment]
@shushanhf This change ported useful `LOONGARCH64` PRs and fixed some errors in `RISCV64`. Thank you so much. And it updates code near `LOONGARCH64`. So could you check to avoid breaking `LOONGARCH64`?

**Comment by shushanhf** [comment]
> @shushanhf This change ported useful `LOONGARCH64` PRs and fixed some errors in `RISCV64`. Thank you so much. And it updates code near `LOONGARCH64`. So could you check to avoid breaking `LOONGARCH64`?

That's ok for LoongArch64.

File: src/coreclr/jit/emitriscv64.cpp [hunk 1]
```diff
@@ -4155,12 +4156,11 @@ regNumber emitter::emitInsTernary(instruction ins, emitAttr attr, GenTree* dst,
         }
         else
         {
-#ifdef DEBUG
             printf("RISCV64-Invalid ins for overflow check: %s\n", codeGen->genInsName(ins));
-#endif
             assert(!"Invalid ins for overflow check");
```
Comment (jakobbotsch) [inline src/coreclr/jit/emitriscv64.cpp:4160]: I think this change should be reverted. The JIT should never print to stdout in release builds, for any reason.

**Review by jakobbotsch — APPROVED** [review]

File: src/coreclr/jit/lsrariscv64.cpp [hunk 2]
```diff
@@ -513,6 +513,11 @@ int LinearScan::BuildNode(GenTree* tree)
         case GT_BOUNDS_CHECK:
         {
             GenTreeBoundsChk* node = tree->AsBoundsChk();
+            if (genActualType(node->GetIndex()->TypeGet()) == TYP_INT)
```
[the remark this answers is not in the document]
Comment (tomeksowi) [reply src/coreclr/jit/lsrariscv64.cpp:516]: thx i will

**Pull request merged**